### PR TITLE
Override ObjC's class_getImageName to handle Swift classes

### PR DESCRIPTION
**Pull request by jrose-apple**
This not only restores the correct behavior for classes with generic ancestry, but also handles actual generic classes as well. (This is the function that backs Foundation's `Bundle.init(for: AnyClass)` initializer.)

Note that this won't actually do anything yet; the ObjC runtime pieces have to be present first, and they aren't. I've run this by @gparker42 already, though, as well as tested it with a custom build of libobjc, and things seem to work.

[SR-1917](https://bugs.swift.org/browse/SR-1917)
rdar://problem/33450609&40367300

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
…That's what I get for adding the arclite change last minute. I'll fix that test in a moment.

File: test/Interpreter/SDK/class_getImageName-static.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  guard let setHookPtr = dlsym(UnsafeMutableRawPointer(bitPattern: -2),
+                               getHookName()) else {
+    // If the version of the ObjC runtime we're using doesn't have the hook,
+    // we're good.
+    return
+  }
+
+  // Note: The following typealias uses AnyObject instead of AnyClass so that
+  // the function type is trivially bridgeable to Objective-C. (The
+  // representation of AnyClass is not the same as Objective-C's 'Class' type.)
```
Comment (jrose-apple) [inline test/Interpreter/SDK/class_getImageName-static.swift]: Oops, and this comment is supposed to be up with the typealias, which I moved.

**Review by mikeash — APPROVED** [review]

**Comment by jrose-apple** [comment]
Test changes only, source compat suite already passed.

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Pull request merged**